packages/metal-component/src/Component.js

```diff
@@ -172,6 +182,8 @@ class Component extends EventEmitter {
 	 */
 	attach(opt_parentElement, opt_siblingElement) {
 		if (!this.inDocument) {
+			this.emit('willAttach');
+			this.willAttach();
```

Comment:
Hey @Robert-Frampton, based on [componentWillMount()](https://reactjs.org/docs/react-component.html#componentwillmount):

> `componentWillMount()` is invoked immediately before mounting occurs. It is called before render(), therefore setting state synchronously in this method will not trigger a re-rendering. Avoid introducing any side-effects or subscriptions in this method.

Can we assume the same with this?

Reply:
Hey @jbalsas 

So an area where Metal is already different than React is that our `render` actually happens before `attach`. This is how the lifecycle is ordered on first render.

`created > willRender > rendered > willAttach > attached > willDetach > detached > disposed`

I think the only way we can mirror React completely is to do a breaking change and change the order so that attachment happens before render, but obviously we can't do that for `2.14`. Although we could move `willAttach` only, but it feels weird to me.

`created > willAttach > willRender > rendered > attached > willDetach > detached > disposed`.

Reply:
That's fine for now... I think we are entitled to some little quirks of our own ;)

I think we just need to improve our [Component Lifecycle](https://metaljs.com/docs/guides/component-lifecycle.html) documentation so everything is clear :)